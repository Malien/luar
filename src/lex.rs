use logos::{Lexer, Logos};
use thiserror::Error;

/// Syntax
///
/// ```
/// module -> { statement | function }
/// block -> { stat sc } [ret sc]
/// sc -> ';'
/// stat -> varlist1 '=' explist1
/// varlist1 -> var { ',' var }
/// var -> name | var '[' exp1 ']' | var '.' name
/// stat -> 'while' exp1 'do' block 'end'
///       | 'repeat' block 'until' exp1
///       | 'if' exp1 'then' block { elseif } ['else' block] 'end'
///       | functioncall
///       | tableconstructor
///       | 'local' declist [init]
/// elseif -> 'elseif' exp1 'then' block
/// ret -> 'return' explist
/// declist -> name { , 'name' }
/// init -> '=' explist1
/// exp -> '(' exp ')'
///      | 'nil'
///      | number
///      | literal
///      | var
/// tableconstructor -> '@' '(' [exp1] ')' | '@' [name] fieldlist
/// fieldlist -> '{' [ffieldlist1] '}' | '[' [lfieldlist1] ']'
/// ffieldlist1 -> ffield { ',' ffield }
/// ffield -> name '=' exp
/// lfieldlist1 -> exp { ',' exp }
/// functioncall -> var '(' [explist1] ')'
/// explist1 -> { exp1 ',' } exp
/// function -> 'function' name '(' [parlist1] ')' block 'end'
/// parlist1 -> 'name' { ',' name }
/// ```

/// Reserved words:
///     `and` `do` `else` `elseif` `end` `function` `if` `local` `nil` `not`
///     `or` `repeat` `return` `until` `then` `while`
///
/// Other tokens:
///     `~=` `<=` `>=` `<` `>` `=` `..` `+` `-` `*` `/` `%` `(` `)` `{` `}` `[` `]` `@` `;` `,` `.`
///
/// Comments denoted by `--` and continue until the end of the line
///
/// Strings are delimited with either single or double quotes, and can contain C-like escape sequences (like `\n` `\r` `\t`)
///
/// Numbers are have either
///     `4` `4.23` `4.` `.23` `4.57e-7` `.3e4`

// #[derive(Copy, Clone, Debug, Eq, PartialEq, Logos)]
// enum Token {
//
// }

#[derive(Clone, Debug, PartialEq, Eq, Logos)]
pub enum Token {
    #[error]
    #[regex(r"[ \t\n\f]", logos::skip)]
    Error,
    #[token("and")]
    And,
    #[token("do")]
    Do,
    #[token("else")]
    Else,
    #[token("elseif")]
    ElseIf,
    #[token("end")]
    End,
    #[token("function")]
    Function,
    #[token("if")]
    If,
    #[token("local")]
    Local,
    #[token("nil")]
    Nil,
    #[token("not")]
    Not,
    #[token("or")]
    Or,
    #[token("repeat")]
    Repeat,
    #[token("return")]
    Return,
    #[token("until")]
    Until,
    #[token("then")]
    Then,
    #[token("while")]
    While,
    #[token("~=")]
    NotEquals,
    #[token("<=")]
    LessOrEquals,
    #[token(">=")]
    GreaterOrEquals,
    #[token(">")]
    Greater,
    #[token("<")]
    Lesser,
    #[token("=")]
    Equals,
    #[token("..")]
    Concat,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Multiply,
    #[token("/")]
    Div,
    #[token("%")]
    Mod,
    #[token("(")]
    OpenRoundBracket,
    #[token(")")]
    CloseRoundBracket,
    #[token("[")]
    OpenSquareBracket,
    #[token("]")]
    CloseSquareBracket,
    #[token("{")]
    OpenSquigglyBracket,
    #[token("}")]
    CloseSquigglyBracket,
    #[token("@")]
    At,
    #[token(".")]
    Dot,
    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[regex(r"[_a-zA-Z][_a-zA-Z0-9]*", make_owned_string)]
    Ident(String),
    #[regex(
        "(\"(?:[^\"'\\\\]|\\\\.)*\")|('(?:[^\"'\\\\]|\\\\.)*')",
        parse_string_literal
    )]
    StringLiteral(String),
    #[regex(
        r"[+-]?((\d+\.\d+)|(\.\d+)|(\d+\.?))(e[+-]?\d+)?",
        parse_number_literal
    )]
    Number(NumberLiteral),
}

#[derive(Error, Debug)]
#[error("Token passed in is not a valid string literal")]
struct StringLiteralParseError;

// enum NumberLiteralParseError {
//     Integral()
// }
type NumberLiteralParseError = num::rational::ParseRatioError;

// Can be optimized not to allocate extra space if not necessary
// enum LuaNumberLiteral {
//     Integral(BigInt),
//     Rational(BigRational)
// }
type NumberLiteral = num::rational::BigRational;

fn parse_string_literal(input: &mut Lexer<Token>) -> Result<String, StringLiteralParseError> {
    let slice = input.slice();
    if slice.len() < 2 {
        return Err(StringLiteralParseError);
    }
    let raw_string: &str = &slice[1..slice.len() - 1];
    // Not a very efficient method to escape characters
    Ok(raw_string
        .replace(r"\\", "\\")
        .replace(r"\r", "\r")
        .replace(r"\n", "\n")
        .replace(r"\t", "\t")
        .replace("\\\"", "\"")
        .replace(r"\'", "'"))
}

// static integer_regex: Regex = Regex::new(r"[+-]?(\d+\.?)(e[+-]?\d+)?").unwrap();

fn parse_number_literal(
    input: &mut Lexer<Token>,
) -> Result<NumberLiteral, NumberLiteralParseError> {
    let slice: &str = input.slice();
    slice.parse()
}

fn make_owned_string(input: &mut Lexer<Token>) -> String {
    input.slice().to_string()
}

#[cfg(test)]
mod tests {
    use super::Token;
    use logos::Logos;

    macro_rules! assert_tokens {
        ($name:ident, $text:tt) => {
            #[test]
            #[function_name::named]
            fn $name() {
                static TEXT: &str = indoc::indoc!($text);
                let tokens: Vec<Token> = Token::lexer(TEXT).into_iter().collect();
                insta::assert_debug_snapshot!(function_name!(), tokens);
            }
        };
    }

    assert_tokens!(function_1, "
        function remove_blanks (s)
            local b = strfind(s, ' ')
            while b do
                s = strsub(s, 1, b-1) .. strsub(s, b+1)
                b = strfind(s, ' ')
            end
            return s
        end
    ");
}
