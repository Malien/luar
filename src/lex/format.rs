use super::Token;

pub struct Formatting {
    pub before: FormattingStyle,
    pub after: FormattingStyle,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FormattingStyle {
    Condensed,
    Space,
    StrictSpace,
    Newline,
    Indent(IndentationChange),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(i8)]
pub enum IndentationChange {
    Increase = 1,
    Decrease = -1,
}

pub fn formatting_style_precedence(
    left: FormattingStyle,
    right: FormattingStyle,
) -> FormattingStyle {
    use FormattingStyle::*;
    use IndentationChange::*;

    match (left, right) {
        (Condensed, Space) => Condensed,
        (Space, Condensed) => Condensed,

        (Condensed, StrictSpace) => StrictSpace,
        (StrictSpace, Condensed) => StrictSpace,

        (Indent(Increase), Indent(Decrease)) => Newline,
        (Indent(c), _) => Indent(c),
        (_, Indent(c)) => Indent(c),

        (Newline, _) => Newline,
        (_, Newline) => Newline,

        // when a == b
        (a, _) => a,
    }
}

impl Token {
    pub fn formatting(&self) -> Formatting {
        use FormattingStyle::*;
        use IndentationChange::*;
        use Token::*;
        match self {
            Error | And | Nil | Not | Or | Equals | NotEquals | LessOrEquals | GreaterOrEquals
            | Greater | Less | Concat | Plus | Minus | Mul | Div | Mod | OpenSquigglyBracket
            | CloseSquigglyBracket | Ident(_) | String(_) | Number(_) => Formatting {
                before: Space,
                after: Space,
            },
            Function | Local => Formatting {
                before: Newline,
                after: Space,
            },
            If | Return | While => Formatting {
                before: Newline,
                after: StrictSpace,
            },
            Else => Formatting {
                before: Indent(Decrease),
                after: Indent(Increase),
            },
            End => Formatting {
                before: Indent(Decrease),
                after: Newline,
            },
            Until | ElseIf => Formatting {
                before: Indent(Decrease),
                after: Space,
            },
            Then | Do | Repeat => Formatting {
                before: Space,
                after: Indent(Increase),
            },
            Exp | OpenRoundBracket | OpenSquareBracket
            | Dot | Colon => Formatting {
                before: Condensed,
                after: Condensed,
            },
            Comma | Semicolon | CloseRoundBracket | CloseSquareBracket => Formatting {
                before: Condensed,
                after: Space,
            },
            Assignment => Formatting {
                before: StrictSpace,
                after: StrictSpace,
            },
        }
    }
}

pub fn format_single_token(
    token: Token,
    indent: &mut i32,
    current_format: &mut FormattingStyle,
    buf: &mut dyn std::fmt::Write,
) -> std::fmt::Result {
    use FormattingStyle::*;
    let Formatting { before, after } = token.formatting();
    match formatting_style_precedence(*current_format, before) {
        Condensed => {}
        Space | StrictSpace => buf.write_char(' ')?,
        Newline => {
            buf.write_char('\n')?;
            for _ in 0..*indent {
                buf.write_char('\t')?;
            }
        }
        Indent(change) => {
            buf.write_char('\n')?;
            *indent += change as i32;
            for _ in 0..*indent {
                buf.write_char('\t')?;
            }
        }
    };
    *current_format = after;
    write!(buf, "{}", token)?;
    Ok(())
}

pub fn format_tokens(
    tokens: &mut impl Iterator<Item = Token>,
    buf: &mut dyn std::fmt::Write,
) -> std::fmt::Result {
    if let Some(first_token) = tokens.next() {
        let mut indent = 0;
        let Formatting {
            after: mut current_format,
            ..
        } = first_token.formatting();
        write!(buf, "{}", first_token)?;

        for token in tokens {
            format_single_token(token, &mut indent, &mut current_format, buf)?;
        }
    }
    Ok(())
}

#[macro_export]
macro_rules! fmt_tokens {
    ($type:ty) => {
        impl std::fmt::Display for $type {
            fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
                crate::lex::format::format_tokens(&mut self.clone().to_tokens(), fmt)
            }
        }
    };
}
