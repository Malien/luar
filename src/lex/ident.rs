use lazy_static::lazy_static;
#[cfg(test)]
use quickcheck::Arbitrary;
use regex::Regex;
use std::{
    iter::{self, FlatMap, Once},
    str::FromStr,
};
use thiserror::Error;

use super::{ToTokenStream, Token};

lazy_static! {
    static ref IDENT_REGEX: Regex = Regex::new(r"^[_a-zA-Z][_a-zA-Z0-9]*$").unwrap();
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Ident(String);

#[derive(Debug, Error)]
#[error("Invalid identifier")]
pub struct InvalidIdentifier;

impl FromStr for Ident {
    type Err = InvalidIdentifier;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if IDENT_REGEX.is_match(s) {
            Ok(Ident(s.to_string()))
        } else {
            Err(InvalidIdentifier)
        }
    }
}

impl Ident {
    #[allow(dead_code)]
    pub fn try_new(str: String) -> Option<Self> {
        if IDENT_REGEX.is_match(&str) {
            Some(Ident(str))
        } else {
            None
        }
    }

    pub fn new(str: impl Into<String>) -> Self {
        Ident(str.into())
    }
}

impl ToTokenStream for Ident {
    type Tokens = iter::Once<Token>;
    fn to_tokens(self) -> Self::Tokens {
        iter::once(Token::Ident(self))
    }
}

impl From<Ident> for String {
    fn from(Ident(str): Ident) -> Self {
        str
    }
}

impl<'a> From<&'a Ident> for &'a String {
    fn from(Ident(str): &'a Ident) -> Self {
        str
    }
}

impl<'a> AsRef<str> for &'a Ident {
    fn as_ref(&self) -> &'a str {
        self.0.as_ref()
    }
}

impl std::fmt::Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)
    }
}

#[cfg(test)]
const VALID_IDENT_CHARS: &'static str =
    "1234567890_abcdefghigklmnopqrstuvwxyzABCDEFGHIGKLMNOPQRSTUVWXYZ";
#[cfg(test)]
const VALID_IDENT_BYTES: &[u8] = VALID_IDENT_CHARS.as_bytes();

#[cfg(test)]
impl Arbitrary for Ident {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        if g.size() == 0 {
            return Ident::new("_");
        }
        let mut buf = Vec::with_capacity(g.size());
        let beginning_bytes = &VALID_IDENT_BYTES[10..];
        // Could use unsafe version unwrap_unchecked()
        buf.push(*g.choose(beginning_bytes).unwrap());
        for _ in 1..g.size() {
            buf.push(*g.choose(VALID_IDENT_BYTES).unwrap());
        }
        // Could use unsafe version from_utf8_unchecked(Vec<u8>)
        let str = String::from_utf8(buf).unwrap();
        Self(str)
    }

    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
        Box::new(self.0.shrink().filter_map(Ident::try_new))
    }
}

impl ToTokenStream for Vec<Ident> {
    // I hate how this is forcing FlatMap to take a function pointer
    // rather than be specialized with exact function type. Gimme impl
    // Iterator, or at least let me name types of specific functions!
    type Tokens = FlatMap<
        std::vec::IntoIter<Ident>,
        Once<Token>,
        fn(Ident) -> <Ident as ToTokenStream>::Tokens,
    >;

    fn to_tokens(self) -> Self::Tokens {
        self.into_iter().flat_map(ToTokenStream::to_tokens)
    }
}
