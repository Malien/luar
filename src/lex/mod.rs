mod ident;
mod number_literal;
mod string_literal;
mod token;

pub type DynTokens = Box<dyn Iterator<Item = Token>>;

pub trait ToTokenStream {
    type Tokens: IntoIterator<Item = Token>;

    fn to_tokens(self) -> Self::Tokens;
}

#[macro_export]
macro_rules! fmt_tokens {
    ($type:ty) => {
        impl std::fmt::Display for $type {
            fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
                for token in self.clone().to_tokens() {
                    write!(fmt, "{} ", token)?;
                }
                Ok(())
            }
        }
    };
}

pub use ident::Ident;
pub use number_literal::NumberLiteral;
pub use string_literal::StringLiteral;
pub use token::Token;
