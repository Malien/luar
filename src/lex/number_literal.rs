use num::pow;
use std::str::{Chars, FromStr};
use thiserror::Error;

#[derive(Debug, Clone, PartialEq, PartialOrd)]
pub enum NumberLiteral {
    Integer(i64),
    Floating(f64),
}

#[derive(Debug, Error)]
#[error("Invalid number format")]
pub struct NumberLiteralParseError;

impl FromStr for NumberLiteral {
    type Err = NumberLiteralParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut chars = s.chars();
        let mut numerator = 0i64;
        let mut sign = 1;
        match chars.next().ok_or(NumberLiteralParseError)? {
            '-' => {
                sign = -1;
            }
            '+' => {}
            '.' => match parse_denumerator(&mut chars)? {
                DenumeratorParsings::Denum(denum) => {
                    return Ok(NumberLiteral::Floating(sign as f64 * denum))
                }
                DenumeratorParsings::DenumWithExponent(denum, exponent) => {
                    let res = if exponent >= 0 {
                        denum * pow(10f64, exponent as usize)
                    } else {
                        denum / pow(10f64, (-exponent) as usize)
                    };
                    return Ok(NumberLiteral::Floating(sign as f64 * res));
                }
            },
            digit_char => {
                if let Some(digit) = digit_char.to_digit(10) {
                    numerator += digit as i64;
                } else {
                    return Err(NumberLiteralParseError);
                }
            }
        };

        loop {
            match chars.next() {
                Some('e') => {
                    let exponent = parse_exponent(&mut chars)?;
                    return Ok(if exponent >= 0 {
                        NumberLiteral::Integer(sign * numerator * pow(10, exponent as usize))
                    } else {
                        NumberLiteral::Floating(
                            ((sign * numerator) as f64) / pow(10f64, (-exponent) as usize),
                        )
                    });
                }
                Some('.') => match parse_denumerator(&mut chars)? {
                    DenumeratorParsings::Denum(denum) => {
                        return Ok(NumberLiteral::Floating(
                            sign as f64 * (numerator as f64 + denum),
                        ))
                    }
                    DenumeratorParsings::DenumWithExponent(denum, exponent) => {
                        let res = if exponent > 0 {
                            (numerator as f64 + denum) * pow(10f64, exponent as usize)
                        } else {
                            (numerator as f64 + denum) / pow(10f64, (-exponent) as usize)
                        };
                        return Ok(NumberLiteral::Floating(sign as f64 * res));
                    }
                },
                Some(digit_char) => {
                    if let Some(digit) = digit_char.to_digit(10) {
                        numerator *= 10;
                        numerator += digit as i64;
                    } else {
                        return Err(NumberLiteralParseError);
                    }
                }
                None => return Ok(NumberLiteral::Integer(numerator * sign)),
            }
        }
    }
}

enum DenumeratorParsings {
    Denum(f64),
    DenumWithExponent(f64, i32),
}

fn parse_denumerator(chars: &mut Chars) -> Result<DenumeratorParsings, NumberLiteralParseError> {
    let mut denumerator = 0f64;
    let mut decade = 10f64;
    loop {
        match chars.next() {
            Some('e') => {
                return Ok(DenumeratorParsings::DenumWithExponent(
                    denumerator,
                    parse_exponent(chars)?,
                ))
            }
            Some(digit_char) => {
                if let Some(digit) = digit_char.to_digit(10) {
                    denumerator += (digit as f64) / decade;
                    decade *= 10f64;
                } else {
                    return Err(NumberLiteralParseError);
                }
            }
            None => return Ok(DenumeratorParsings::Denum(denumerator)),
        }
    }
}

fn parse_exponent(chars: &mut Chars) -> Result<i32, NumberLiteralParseError> {
    let mut exponent = 0;
    let mut sign = 1;
    loop {
        match chars.next() {
            Some('+') => {}
            Some('-') => {
                sign = -1;
            }
            Some(digit_char) => {
                if let Some(digit) = digit_char.to_digit(10) {
                    exponent *= 10;
                    exponent += digit as i32;
                } else {
                    return Err(NumberLiteralParseError);
                }
            }
            None => return Ok(exponent * sign),
        }
    }
}

#[cfg(test)]
mod tests {
    use num::{abs, pow};
    use quickcheck::TestResult;

    use super::NumberLiteral;

    #[quickcheck]
    fn parses_integers(input: i64) {
        let res: NumberLiteral = input.to_string().parse().unwrap();
        assert_eq!(NumberLiteral::Integer(input), res);
    }

    #[quickcheck]
    fn parses_with_dot(input: i64) {
        let res: NumberLiteral = format!("{}.", input).parse().unwrap();
        assert_eq!(NumberLiteral::Floating(input as f64), res);
    }

    #[quickcheck]
    fn parses_with_dot_before(input: u64) {
        let res: NumberLiteral = format!(".{}", input).parse().unwrap();
        let expected: f64 = format!("0.{}", input).parse().unwrap();
        assert_eq_f64(expected, res);

        let res: NumberLiteral = format!("-.{}", input).parse().unwrap();
        let expected: f64 = format!("-0.{}", input).parse().unwrap();
        assert_eq_f64(expected, res);
    }

    #[quickcheck]
    fn parses_floats(input: f64) {
        let res: NumberLiteral = input.to_string().parse().unwrap();
        assert_eq_f64(input, res);
    }

    #[quickcheck]
    fn parses_integers_with_exponent(input: i64, exponent: u8) -> TestResult {
        // TODO: fix overflows with exponents and/or long numbers
        if exponent > 16 {
            return TestResult::discard();
        }

        let res: NumberLiteral = format!("{}e{}", input, exponent).parse().unwrap();
        let expected = input * pow(10, exponent as usize);
        assert_eq!(NumberLiteral::Integer(expected), res);

        if exponent == 0 {
            return TestResult::passed()
        }

        let res: NumberLiteral = format!("{}e-{}", input, exponent).parse().unwrap();
        let expected = input as f64 / pow(10f64, exponent as usize);
        assert_eq_f64(expected, res);

        TestResult::passed()
    }

    #[quickcheck]
    fn parses_integers_with_negative_zero_exponent(input: i64) {
        let res: NumberLiteral = format!("{}e-0", input).parse().unwrap();
        assert_eq!(NumberLiteral::Integer(input), res);
    }

    #[quickcheck]
    fn parses_with_dot_with_exponent(input: i64, exponent: u8) -> TestResult {
        if exponent > 16 {
            return TestResult::discard();
        }

        let res: NumberLiteral = format!("{}.e{}", input, exponent).parse().unwrap();
        let expected = input * pow(10, exponent as usize);
        assert_eq!(NumberLiteral::Floating(expected as f64), res);

        let res: NumberLiteral = format!("{}.e-{}", input, exponent).parse().unwrap();
        let expected = input as f64 / pow(10f64, exponent as usize);
        assert_eq_f64(expected, res);

        TestResult::passed()
    }

    #[quickcheck]
    fn parses_with_dot_before_with_exponent(input: u64, exponent: u8) -> TestResult {
        if exponent > 16 {
            return TestResult::discard();
        }

        let base_number: f64 = format!("0.{}", input).parse().unwrap();

        let res: NumberLiteral = format!(".{}e{}", input, exponent).parse().unwrap();
        assert_eq_f64_with_exponent(base_number, res, exponent);

        let res: NumberLiteral = format!("-.{}e{}", input, exponent).parse().unwrap();
        assert_eq_f64_with_exponent(-base_number, res, exponent);

        let res: NumberLiteral = format!(".{}e-{}", input, exponent).parse().unwrap();
        let expected = base_number / pow(10f64, exponent as usize);
        // TODO: reverse exponent
        assert_eq_f64(expected, res);

        let res: NumberLiteral = format!("-.{}e-{}", input, exponent).parse().unwrap();
        let expected: f64 = -base_number / pow(10f64, exponent as usize);
        // TODO: reverse exponent
        assert_eq_f64(expected, res);

        TestResult::passed()
    }

    #[quickcheck]
    fn parses_floats_with_exponent(input: f64, exponent: u8) -> TestResult {
        if exponent > 16 {
            return TestResult::discard();
        }

        let res: NumberLiteral = format!("{}e{}", input, exponent).parse().unwrap();
        assert_eq_f64_with_exponent(input, res, exponent);

        TestResult::passed()
    }

    fn assert_eq_f64(expected: f64, got: NumberLiteral) {
        if let NumberLiteral::Floating(got) = got {
            assert!(
                close_cmp(got, expected, 0.0000001),
                "Expected {:?}, got {:?}",
                NumberLiteral::Floating(expected),
                got
            );
        } else {
            panic!("Expected {:?} to be NumberLiteral::Floating", got);
        }
    }

    fn assert_eq_f64_with_exponent(expected: f64, got: NumberLiteral, exponent: u8) {
        if let NumberLiteral::Floating(got) = got {
            let decade = pow(10f64, exponent as usize);
            assert!(
                close_cmp(got, expected * decade, decade * 0.0000001),
                "Expected {:?}, got {:?}",
                NumberLiteral::Floating(expected),
                got
            );
        } else {
            panic!("Expected {:?} to be NumberLiteral::Floating", got);
        }
    }

    fn close_cmp(a: f64, b: f64, eps: f64) -> bool {
        abs(a - b) < eps
    }
}
