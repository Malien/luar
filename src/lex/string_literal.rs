use std::str::FromStr;

use thiserror::Error;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct StringLiteral(String);

#[derive(Error, Debug)]
#[error("Token passed in is not a valid string literal")]
pub struct StringLiteralParseError;

impl FromStr for StringLiteral {
    type Err = StringLiteralParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.len() < 2 {
            return Err(StringLiteralParseError);
        }
        let raw_string: &str = &s[1..s.len() - 1];
        // Not a very efficient method to escape characters
        Ok(StringLiteral(
            raw_string
                .replace(r"\\", "\\")
                .replace(r"\r", "\r")
                .replace(r"\n", "\n")
                .replace(r"\t", "\t")
                .replace("\\\"", "\"")
                .replace(r"\'", "'"),
        ))
    }
}
// fn parse_string_literal(input: &mut Lexer<Token>) -> Result<String, StringLiteralParseError> {
//     let slice = input.slice();
// }
