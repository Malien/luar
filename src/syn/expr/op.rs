use std::iter;

use crate::lex::{ToTokenStream, Token};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum BinaryOperator {
    // Precedence level 0
    And,
    Or,
    // Precedence level 1
    Less,
    Greater,
    LessOrEquals,
    GreaterOrEquals,
    NotEquals,
    Equals,
    // Precedence level 2
    Concat,
    // Precedence level 3
    Plus,
    Minus,
    // Precedence level 4
    Mul,
    Div,
    // Precedence level 5
    Exp,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum UnaryOperator {
    Minus,
    Not,
}

impl ToTokenStream for BinaryOperator {
    type Tokens = iter::Once<Token>;
    fn to_tokens(self) -> Self::Tokens {
        iter::once(match self {
            BinaryOperator::And => Token::And,
            BinaryOperator::Or => Token::Or,
            BinaryOperator::Less => Token::Less,
            BinaryOperator::Greater => Token::Greater,
            BinaryOperator::LessOrEquals => Token::LessOrEquals,
            BinaryOperator::GreaterOrEquals => Token::GreaterOrEquals,
            BinaryOperator::NotEquals => Token::NotEquals,
            BinaryOperator::Equals => Token::Equals,
            BinaryOperator::Concat => Token::Concat,
            BinaryOperator::Plus => Token::Plus,
            BinaryOperator::Minus => Token::Minus,
            BinaryOperator::Mul => Token::Mul,
            BinaryOperator::Div => Token::Div,
            BinaryOperator::Exp => Token::Exp,
        })
    }
}

impl ToTokenStream for UnaryOperator {
    type Tokens = iter::Once<Token>;
    fn to_tokens(self) -> Self::Tokens {
        iter::once(match self {
            UnaryOperator::Minus => Token::Minus,
            UnaryOperator::Not => Token::Not,
        })
    }
}

#[cfg(test)]
pub mod test {
    use super::{BinaryOperator, UnaryOperator};
    use quickcheck::{Arbitrary, Gen};

    impl Arbitrary for UnaryOperator {
        fn arbitrary(g: &mut Gen) -> Self {
            match u8::arbitrary(g) % 2 {
                0 => UnaryOperator::Minus,
                1 => UnaryOperator::Not,
                _ => unreachable!(),
            }
        }
    }

    impl Arbitrary for BinaryOperator {
        fn arbitrary(g: &mut Gen) -> Self {
            match u8::arbitrary(g) % 14 {
                0 => BinaryOperator::And,
                1 => BinaryOperator::Or,
                2 => BinaryOperator::Less,
                3 => BinaryOperator::Greater,
                4 => BinaryOperator::LessOrEquals,
                5 => BinaryOperator::GreaterOrEquals,
                6 => BinaryOperator::NotEquals,
                7 => BinaryOperator::Equals,
                8 => BinaryOperator::Concat,
                9 => BinaryOperator::Plus,
                10 => BinaryOperator::Minus,
                11 => BinaryOperator::Mul,
                12 => BinaryOperator::Div,
                13 => BinaryOperator::Exp,
                _ => unreachable!(),
            }
        }
    }
}
