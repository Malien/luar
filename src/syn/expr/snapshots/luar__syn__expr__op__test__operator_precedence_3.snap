---
source: src/syn/expr/op.rs
expression: expression
---
BinaryOperator {
    lhs: BinaryOperator {
        lhs: Number(
            NumberLiteral(
                1.0,
            ),
        ),
        op: LessOrEquals,
        rhs: BinaryOperator {
            lhs: BinaryOperator {
                lhs: BinaryOperator {
                    lhs: Number(
                        NumberLiteral(
                            2.0,
                        ),
                    ),
                    op: And,
                    rhs: Number(
                        NumberLiteral(
                            3.0,
                        ),
                    ),
                },
                op: Plus,
                rhs: UnaryOperator {
                    op: Not,
                    exp: BinaryOperator {
                        lhs: Number(
                            NumberLiteral(
                                4.0,
                            ),
                        ),
                        op: Mul,
                        rhs: Number(
                            NumberLiteral(
                                5.0,
                            ),
                        ),
                    },
                },
            },
            op: Exp,
            rhs: Number(
                NumberLiteral(
                    6.0,
                ),
            ),
        },
    },
    op: GreaterOrEquals,
    rhs: BinaryOperator {
        lhs: BinaryOperator {
            lhs: Number(
                NumberLiteral(
                    7.0,
                ),
            ),
            op: Minus,
            rhs: BinaryOperator {
                lhs: UnaryOperator {
                    op: Minus,
                    exp: Number(
                        NumberLiteral(
                            8.0,
                        ),
                    ),
                },
                op: Div,
                rhs: BinaryOperator {
                    lhs: BinaryOperator {
                        lhs: Number(
                            NumberLiteral(
                                9.0,
                            ),
                        ),
                        op: Or,
                        rhs: Number(
                            NumberLiteral(
                                10.0,
                            ),
                        ),
                    },
                    op: NotEquals,
                    rhs: Number(
                        NumberLiteral(
                            11.0,
                        ),
                    ),
                },
            },
        },
        op: Concat,
        rhs: Number(
            NumberLiteral(
                12.0,
            ),
        ),
    },
}
